16-5

KinCadeFoster opened the pull request:
- создание классов и методов для героя игры

KinCadeFoster committed:
16-5
- создание классов и методов для героя игры

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    fun healing(heal: Int) {
+        if (health > 0) health += heal
```
Предлагаю добавить проверку на ограничение максимального здоровья, чтобы не превысить лимит здоровья.

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,45 @@
+) {
+
+    fun takingDamage(damage: Int) {
+        health -= damage
```
Следует вычитать из здоровья абсолютное значение дамага (есть функция).

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,45 @@
+    fun takingDamage(damage: Int) {
+        health -= damage
+        if (health < 1) die()
+        println("Ваш герой получил $damage едениц урона")
```
Нет смысла выводить это сообщение если герой уже погиб.

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,45 @@
+
+    fun healing(heal: Int) {
+        if (health > 0) health += heal
+        println("Ваш герой восполнил $heal едениц урона")
```
Нет смысла выводить это сообщение если герой уже погиб.

KinCadeFoster committed:
16-5
- рефакторинг

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,62 @@
+    fun takingDamage(damage: Int) {
+        health -= abs(damage)
+        if (health < 1)  {
+            println("Ваш герой получил $damage едениц урона.") // Игрок же должен понять от чего погиб)
```
Все верно, он должен узнать, однако он же продолжит получать сообщение даже когда уже будет мертв, а  это не правильно. Это можно  решить либо переделав более хитро условия, либо добавив флаг isDead.

KinCadeFoster committed:
16-5
- рефакторинг

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    fun healing(heal: Int) {
+        if (health == 100) {
```
Магическое число 100 стоит вынести как поле класса, так как ни где не указано что максимальное здоровье это 100, и оно много где используется, что усложнит любые правки этого параметра в будущем.

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,65 @@
+                    "Максимальное здровье не может быть более 100 едениц!")
+            health = 100
+        }
+        else if (health + heal in 1..99){
```
in не лучше решения для сравнения больше/меньше.

KinCadeFoster committed:
16-5
- вынес максимальноездровье в константу
- оптимизировал часть условии
- поправил баг, после предыдущих правок герой не получал урон, если урон не убивал его)

ievetrovdev commented on src/main/kotlin/lesson_16/lesson16_task5.kt:
```diff
@@ -0,0 +1,70 @@
+
+import kotlin.math.abs
+
+private const val MAX_HP = 100
```
Вынести нужно как поле класса, все таки максимальное здоровье это не константа для всех, наверняка разные герои (экземпляры класса) могут иметь разное максимальное здоровье, которое они получают при создании.

KinCadeFoster committed:
16-5
- вынес  константу в поле класса maxHealth

ievetrovdev reviewed (approved):
Теперь все отлично, прекрасная работа!
